service/frontend/workflow_handler.go

```diff
@@ -3610,9 +3610,14 @@ func (wh *WorkflowHandler) ListBatchOperations(
 		return nil, errBatchAPINotAllowed
 	}
 
+	maxPageSize := int32(wh.config.VisibilityMaxPageSize(request.GetNamespace()))
+	if request.GetPageSize() <= 0 || request.GetPageSize() > maxPageSize {
+		request.PageSize = maxPageSize
+	}
+
 	resp, err := wh.ListWorkflowExecutions(ctx, &workflowservice.ListWorkflowExecutionsRequest{
 		Namespace:     request.GetNamespace(),
-		PageSize:      int32(wh.config.VisibilityMaxPageSize(request.GetNamespace())),
+		PageSize:      request.GetPageSize(),
```

Comment:
You don't need to call this as it just returns the page size directly.
```suggestion
		PageSize:      request.PageSize,
```

You could also do this using a local pageSize var, instead of updating the request struct

Reply:
That's all true. I opted for this approach because of consistency; I saw a lot of places in this file do it this way so I followed. But I can commit the suggestion since it's objectively more efficient :)